### PR TITLE
DCAC-47 - Fresh PR for handling POST request

### PR DESCRIPTION
**Pull request by rmcmurray-ch**

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java [hunk 1]
```diff
@@ -1,5 +1,7 @@
 package uk.gov.companieshouse.itemgroupworkflowapi.controller;
 
+import org.springframework.boot.test.context.TestConfiguration;
```
Comment (caismith-ch) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java:3]: This looks like an unused import

File: src/test/java/uk/gov/companieshouse/itemgroupworkflowapi/integration/ItemGroupControllerIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    private ItemGroupsRepository repository;
+
+    @AfterEach
+    void tearDown() {
+        repository.deleteAll();
+    }
+
+    @Test
+    @DisplayName("Create successful itemGroup - 201 Created")
+    void createItemGroupSuccessful_201Created() throws Exception {
```
Comment (caismith-ch) [inline src/test/java/uk/gov/companieshouse/itemgroupworkflowapi/integration/ItemGroupControllerIntegrationTest.java:60]: This test is currently failing and the test method name should be updated to remove the underscore to be `createItemGroupSuccessful_201Created`

File: src/test/java/uk/gov/companieshouse/itemgroupworkflowapi/integration/ItemGroupControllerIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .accept(MediaType.APPLICATION_JSON)
+                        .content(mapper.writeValueAsString(newItemGroupData)))
+                .andExpect(status().isCreated())
+                .andDo(MockMvcResultHandlers.print());
+    }
+
+    @Test
+    @DisplayName("Create itemGroup unsuccessful - 400 Bad Request")
+    void createItemGroupUnsuccessful_400BadRequest() throws Exception {
```
Comment (caismith-ch) [inline src/test/java/uk/gov/companieshouse/itemgroupworkflowapi/integration/ItemGroupControllerIntegrationTest.java:77]: This test is currently failing and the test method name should be updated to remove the underscore to be `createItemGroupUnsuccessful400BadRequest`

File: src/test/java/uk/gov/companieshouse/itemgroupworkflowapi/integration/ItemGroupControllerIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .accept(MediaType.APPLICATION_JSON)
+                        .content(mapper.writeValueAsString(newItemGroupData)))
+                .andExpect(status().isBadRequest())
+                .andDo(MockMvcResultHandlers.print());
+    }
+
+    @Test
+    @DisplayName("create duplicate itemGroup fails - 409 Conflict")
+    void createDuplicateItemGroupUnsuccessful_409Conflict() throws Exception {
```
Comment (caismith-ch) [inline src/test/java/uk/gov/companieshouse/itemgroupworkflowapi/integration/ItemGroupControllerIntegrationTest.java:94]: This test is currently failing and the test method name should be updated to remove the underscore to be `createDuplicateItemGroupUnsuccessful_409Conflict`

File: src/test/postman/item-group-workflow-api.postman_collection.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
 	"info": {
-		"_postman_id": "bc84bc92-bdd8-4cb6-8242-7192a45a6798",
-		"name": "item-group-workflow-api",
+		"_postman_id": "bb7c75a6-8959-42dc-ab34-611f5318b3c2",
+		"name": "item-group-workflow-api Copy 2",
```
Comment (caismith-ch) [inline src/test/postman/item-group-workflow-api.postman_collection.json:4]: What does the `Copy 2` refer to, is it required?

File: pom.xml [hunk 6]
```diff
@@ -109,6 +116,38 @@
...
+            <artifactId>spring-boot-test</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>${guava.version}</version>
+        </dependency>
+
+        <dependency>
```
Comment (VaughanJackson123) [inline pom.xml:130]: I am not sure we need all of these `jakarta.json` etc dependencies? They are probably only required for the PATCH implementation?

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java [hunk 7]
```diff
@@ -13,96 +15,118 @@
...
+                return buildItemAlreadyExistsResponse(xRequestId, itemGroupData);
 
             final ItemGroup savedItem = itemGroupsService.createItemGroup(itemGroupData);
-            return buildCreateSuccessResponse(savedItem);
-
-        } catch (Exception ex) {
-            return buildErrorResponse(ex, itemGroupData);
+            return buildCreateSuccessResponse(xRequestId, savedItem);
+        }
+        catch (Exception ex) {
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java:66]: Are we planning on dealing with exceptions in a finer-grained way in another PR then?

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/validation/ItemGroupsValidator.java [hunk 8]
```diff
@@ -1,97 +1,142 @@
...
+    public static final String COMPANY_NUMBER_MISSING = "Items Group Delivery Details : missing item company number for : ";
+    public static final String COMPANY_NAME_MISSING = "Items Group Delivery Details : missing item company name for : ";
+
+    public List<String> validateCreateItemData(ItemGroupData dto) {
+        final List<String> errors = new ArrayList<>();
 
-    private final ItemGroupsService itemGroupsService;
+        if (orderNumberValid(dto, errors) &&
+            itemsPresent(dto, errors) &&
+            linksPresentWithOrderNumber(dto, errors))
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/validation/ItemGroupsValidator.java:36]: I think we need to validate the links on the items too. Perhaps in a separate PR/ticket?

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/service/ItemGroupsService.java [hunk 9]
```diff
@@ -34,6 +47,8 @@ public ItemGroup createItemGroup(ItemGroupData itemGroupData) {
         setCreationTimeStamp(itemGroup);
         itemGroup.setData(itemGroupData);
 
+        regenerateLinks(itemGroupData, itemGroupId);
+
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/service/ItemGroupsService.java:51]: I think in a way similar (but simpler) to the way we generate the `self` links here on persisting the item group, we will want to generate an `etag` too. This was recently added to the spec in [DCAC-53](https://github.com/companieshouse/private.api.ch.gov.uk-specifications/pull/158). Perhaps this could also be dealt with in a separate PR/ticket?

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/validation/ItemGroupsValidator.java (hunk 8 above)
Comment (caismith-ch) [reply src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/validation/ItemGroupsValidator.java:36]: I'll speak to Frank about getting some placeholder tickets added that we can refine and bring in potentially during mid iteration.

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java (hunk 7 above)
Comment (rmcmurray-ch) [reply src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java:66]: Yep, we are ready to add exception handling to deal with internal server errors etc, but potentially under another ticket that deals with improvements to the current exception handling so that dcac-47 can be completed 👍

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/validation/ItemGroupsValidator.java (hunk 8 above)
Comment (caismith-ch) [reply src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/validation/ItemGroupsValidator.java:36]: `DCAC-247` and `DCAC-248` have been raised and moved to the top of the backlog to `Validate item links in item-group-workflow-api` and `Expand ItemGroupController exception handling`

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/service/ItemGroupsService.java (hunk 9 above)
Comment (rmcmurray-ch) [reply src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/service/ItemGroupsService.java:51]: Good idea 👍  @caismith-ch could this be included with the placeholder tickets?

**Comment by rmcmurray-ch** [comment]
@VaughanJackson123 @caismith-ch latest comments have been addressed in last commit 👍

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/service/ItemGroupsService.java (hunk 9 above)
Comment (caismith-ch) [reply src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/service/ItemGroupsService.java:51]: Yes definitely, will get it raised now

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java [hunk 10]
```diff
@@ -13,96 +14,118 @@
...
         if (!errors.isEmpty()) {
-            return buildValidationResponse(errors, itemGroupData);
+            return buildValidationResponse(xRequestId, errors);
         }
 
         try {
-            if (itemGroupsService.doesItemGroupExist(itemGroupData)) {
-                return buildItemAlreadyExistsResponse(itemGroupData);
-            }
+            if (itemGroupsService.doesItemGroupExist(itemGroupData))
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java:59]: The way in which `doesItemGroupExist` is derived will need to change, [as discussed in Slack](https://companieshouse.slack.com/archives/C02TZ3RCGG5/p1687268405949549). I think this is probably another candidate for a separate PR/ticket. We should also confirm what behaviour is expected with the architects.

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/service/ItemGroupsService.java (hunk 9 above)
Comment (caismith-ch) [reply src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/service/ItemGroupsService.java:51]: Raised under `DCAC-249`

**Review by caismith-ch — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java [hunk 11]
```diff
@@ -13,96 +14,118 @@
...
+    private ResponseEntity<Object> buildCreateSuccessResponse(String xRequestId,
+                                                              final ItemGroup savedItem) {
         DataMap dataMap = new DataMap.Builder()
+            .requestId(xRequestId)
             .orderId(savedItem.getData().getOrderNumber())
             .build();
 
-        logger.getLogger().info(CREATE_ITEM_GROUP_CREATED, dataMap.getLogMap());
+        logger.getLogger().info(CREATE_ITEM_GROUP_CREATED_PREFIX, dataMap.getLogMap());
         return ResponseEntity.status(CREATED).body(savedItem);
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java:88]: As previously discussed, this should probably be populating the response payload with the nested `ItemGroupData` object rather than its nesting `ItemGroup` object, to avoid returning metadata in the response.

Comment (VaughanJackson123) [reply src/main/java/uk/gov/companieshouse/itemgroupworkflowapi/controller/ItemGroupController.java:88]: Looks good to me. 👍
